### PR TITLE
Introduce new plugin: local pseudo plugin

### PR DESCRIPTION
**Pull request by cathay4t**
* Just a wrapper of hpsa/megaraid/arcconf which save users from guessing
   which plugin they should use.
 * Use kernel module to determine which plugin to load:
```
        _KMOD_PLUGIN_MAP = {
            'megaraid_sas': 'megaraid',
            'smartpqi': 'arcconf',
            'hpsa': 'hpsa',
            'aacraid': 'arcconf',
            'nfsd': 'nfs',
        }
```
 * Manpage included.
 * Tested on a server with both SmartArray and MegaRAID, along with NFS services.
 * URI parameters:

   * `ignore_init_error=true|false`, default is raise all errors.
   * `only=<plugin_name>` -- only load certain plugin.
   * `<plugin_name>_<uri_parameter_name>=<value>` -- passing parameters to certain plugin.

### CODE REVIEW

**Comment by cathay4t** [comment]
@joehandzik @brraghav If possible, please test this plugin in your lab. The command to try is:
`lsmcli ls -u 'local://'`. Thank you.

**Comment by brraghav** [comment]
Looks like a good idea ! 
But I see some issue in case of arcconf. In case of Adaptec Series 6/7/8 controllers, the approach would work. In case of Smart array controllers, both arcconf and hpsa use smartpqi driver. We need to handle this situation. The reason is, both arcconf and ssacli co-exist. 
Will test this command on Adaptec series 8 controllers and get back.

**Comment by cathay4t** [comment]
@joehandzik I noticed in Linux kernel, `smartpqi` is named as `Driver for Microsemi Smart Family Controller`. Hence I intended to mark `smartpqi` handled by `arcconf` plugin, is that OK for you?

@brraghav Could test local plugin with `local://?only=arcconf` on smartpqi hardwares?

Thank you.

**Comment by cathay4t** [comment]
Let me find a way better way to handle smartpqi here.

**Comment by joehandzik** [comment]
@cathay4t I think the right way to handle this would be to find some way to make it possible for both plugins to work. Unfortunately, HPE will continue to use a different tool than the Adaptec Controllers, so there's no clear way to distinguish the two cases by driver alone. I'll give this a try when I have a moment!

**Comment by cathay4t** [comment]
* For smartpqi kernel module, since both hpsa and arcconf is supported. 
   We still use the plugin which binary tool is installed(try arcconf
   first).
 * Trivial changed to arcconf and hpsa plugin to expose function which
   searching binary tool.

File: doc/man/local_lsmplugin.1.in [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+.SH NAME
+local_lsmplugin -- LibstorageMgmt Local Pseudo plugin
+
+.SH DESCRIPTION
+LibstorageMgmt local plugin allows user to manage locally attached storage
+hardware/software without caring which real plugin(s) should be used.
+The 'local_lsmplugin' executable file is for libStorageMgmt
+daemon to execute when client user specifies local plugin in the URI.
+
+Currently, local plugin is wrapping of these plugins:
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: Currently, local plugin wraps these plugins:

File: doc/man/local_lsmplugin.1.in [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+
+    * MegaRAID plugin -- `megariad_lsmplugin(1)`.
+    * SmartArray plugin -- `hpsa_lsmplugin(1)`.
+    * Adaptec plugin -- `arcconf_lsmplugin(1)`.
+
+Certain binrary tool might be required, please check error message or manpage
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: Specific binary tool may be required,

File: doc/man/local_lsmplugin.1.in [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+of above plugins.
+
+Extra SELinux actions are required to allowing plugin access the
+hardware/software.
+
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: Additional SELinux actions are required to allow plugin access to hardware/software.

**Note: Can we address this so the users don't need to intervene?**

File: doc/man/local_lsmplugin.1.in [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+.TP
+\fBURI parameters\fR
+These URI parameters are supported by this plugin:
+
+.RS 7
+.TP
+\fBignore_init_error=true\fR
+
+The 'ignore_init_error' URI parameter is used to ignore errors when initialing
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: typo: `initializing`

File: doc/man/local_lsmplugin.1.in [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+real plugins. But if no valid system been found, an error will still be raised.
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: If no valid systems are found, an error will still be raised.

File: doc/man/local_lsmplugin.1.in [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+By default, all errors will be raised.
+
+.TP
+\fBonly=<plugin_name>\fR
+
+The 'only' URI parameter is used to force local plugin only load certain
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: ... force local plugin to only load specific real plugin(s).

File: doc/man/local_lsmplugin.1.in [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+real plugins.
+
+By default, all plugins will be loaded if supported hardware have been found.
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: has been found.

File: doc/man/local_lsmplugin.1.in [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+.TP
+\fB<plugin_name>_<uri_parameter_name>=<value>\fR
+
+This URI parameter format is used to pass URI parameters to real plugin.
+For example, to pass 'storcli=/usr/bin/storcli' URI parameter to MegaRAID plugin,
+you should use 'megaraid_storcli=/usr/bin/storcli'.
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: you would use

File: doc/man/local_lsmplugin.1.in [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+.SH ROOT PRIVILEGE
+This plugin requires both \fBlsmd\fR daemon and API client running as root
+user. Please check manpage \fIlsmd.conf (5)\fR for detail.
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: details.

File: doc/man/local_lsmplugin.1.in [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    * Adaptec plugin -- `arcconf_lsmplugin(1)`.
+
+Detailed support status can be queried via:
+
+ * \fBlsm.Client.capabilities()\fR  (Python API)
+ * \fBlsm_capabilities()\fR         (C API)
+ * \fBlsmcli capabilities\fR        (lsmcli command line).
+
+.SH FIREWALL RULES
+This plugin only execute tools on localhost. No network connection required.
```
Comment (tasleson) [inline doc/man/local_lsmplugin.1.in]: executes

File: doc/man/local_lsmplugin.1.in (hunk 3 above)
Comment (joehandzik) [inline doc/man/local_lsmplugin.1.in]: Wouldn't this be a problem still with the sub-local plugins themselves too? I didn't see mention of SELinux on the existing HP manpage, for example.

File: plugin/local/__init__.py [hunk 11]
```diff
@@ -0,0 +1 @@
+from lsm.plugin.megaraid.megaraid import MegaRAID
```
Comment (joehandzik) [inline plugin/local/__init__.py]: Why is this import here in the __init__.py, while the other two are in plugin/local/local.py?

**Comment by joehandzik** [comment]
@cathay4t This is really awesome! Added my thoughts alongside @tasleson's. I did not have a chance to test this yet on an HPE box, but I'll test against both a Gen9 and Gen10 controller so we get a look at both hpsa and smartpqi environments. Out of curiosity, have you been able to test a box with multiple controller vendors present? I don't have access to a configuration like that yet.

**Comment by joehandzik** [comment]
@cathay4t Just read your commit message, it seems you did test this on a box with both vendors at the same time. If that worked as expected, this satisfies my requirements from issue #229.

**Comment by cathay4t** [comment]
@joehandzik I have a HPE server with both MegaRAID and SmartArray Gen 9 installed. It works well. I might have a smartpqi card soon, but not sure when.

File: plugin/local/__init__.py (hunk 11 above)
Comment (cathay4t) [inline plugin/local/__init__.py]: I just copy it from MegaRAID plugin, will remove it. Thanks.

File: doc/man/local_lsmplugin.1.in (hunk 3 above)
Comment (cathay4t) [inline doc/man/local_lsmplugin.1.in]: I believe I should remove that line. LSM works well on Fedora and RHEL. That was an old line when SELinux block external binrary tools of storcli and hpssacli.

**Comment by cathay4t** [comment]
The CI test failure of `ppc64_make_check_rh_lab` is caused by #284 

Man page and `__init__` file fixed.
Also fixed the `__del__` function.

**Comment by joehandzik** [comment]
@cathay4t @tasleson Just tested on my boxes, everything looks good!

**Comment by cathay4t** [comment]
Please hold on, I am adding support of NFS plugin to local plugin.

**Comment by cathay4t** [comment]
Rebased to current master and also included the support of NFS plugin which is determined by kernel module `nfsd`.

**Comment by tasleson** [comment]
@cathay4t  Thanks Gris!
